site/tags/server-sidebar.tag

```diff
@@ -0,0 +1,647 @@
+
+      this.update()
+    })
+
+    RiotControl.on('session_id_update', async sessionID => {
+      this.loggedIn = '?'
+      const sessionObj = sessionID ? await get(this.currentServerURL, 'session/' + sessionID) : { success: false }
+
+      if (sessionObj.success) {
+        this.loggedIn = true
+        this.username = sessionObj.user.username
+      } else {
+        this.loggedIn = false
+        this.username = undefined
+      }
+
+      this.update()
+    })
+
+    showLoginModal(evt) {
+      evt.preventDefault()
+      this.refs.loginModal.open()
+    }
+
+    async submitLoginModal({ username, password }) {
+      this.loginInProgress = true
+
+      const result = await post(this.currentServerURL, 'login', { username, password })
+
+      if (result.error) {
+        if (result.error === 'user not found') {
+          throw `There is no user with the username ${username}.`
+        } else if (result.error === 'incorrect password') {
+          throw 'Incorrect password!'
+        } else {
+          // Unknown error :/
+          console.error('Unknown error while logging in', result.error)
+          throw result.error
+        }
+      } else {
+        // Success!!
+        RiotControl.trigger('session_id_update', result.sessionID)
+      }
+    }
+
+    showRegisterModal(evt) {
+      evt.preventDefault()
+      this.refs.registerModal.open()
+    }
+
+    async submitRegisterModal({ username, password }) {
+      this.loginInProgress = true
+
+      const result = await post(this.currentServerURL, 'register', { username, password })
+
+      if (result.error) {
+        if (result.error === 'username already taken') {
+          throw `Username '${username}' already taken.`
+        } else if (result.error === 'username invalid') {
+          throw `Username '${username}' is invalid`
+        } else if (result.error === 'password must be at least 6 characters long') {
+          throw 'Password must be at least 6 characters long'
+        } else {
+          // Unknown error :/
+          console.error('Unknown error while registering', result.error)
+          throw result.error
+        }
+      } else if (result.success) {
+        // Success!!
+      }
+    }
+
+    logout() {
+      RiotControl.trigger('session_id_update', null)
+    }
+
+    showAddServerModal(evt) {
+      evt.preventDefault()
+      this.refs.addServerModal.open()
+    }
+
+    async submitAddServerModal({ url }) {
+      try {
+        const res = await fetch('//' + url + '/api/')
+
+        // The status code returned should be 418
+        if (res.status !== 418) {
+          throw -1
+        }
+
+        // We should see { bantisocial: true }
+        const { bantisocial } = await res.json()
+        if (bantisocial !== true) {
+          throw -1
+        }
+      } catch (err) {
+        if (err !== -1) {
+          console.warn('Error whilst adding server', err)
+        }
+
+        throw 'Not a bantisocial chat server'
```

Comment:
Note that this will be thrown *even if* the error code is not -1. You might want to add a `throw 'Error adding server (see browser console)'` line to `if (err !== -1)`?

Reply:
I figured it's better to keep the user out of the error and let devs handle it..?

Reply:
Mm. But it's not *not* a bantisocial server - after all, it did respond with the correct metadata and such. A more general error message (even just "Error adding server") should be better, so we aren't lying/confusing users.